### ISSUE TITLE
Send linking updates and deletes to DCP

### ISSUE THREAD

**Issue by aaclan-ebi** [ebi-ait/dcp-ingest-central#695]
## User story
As a data submitter
I want to submit linking updates to DCP
so that I can submit the correct model of my experiment 

## Acceptance criteria 
Some are already supported with the current implementation but adding them again here so we can test them again as we're changing the design/implementation. Criteria 1, 3, 4  are already supported and functional as implemented from dcp-680.

#### no new assay 
1.  Given 2 donors and cell suspension and cell suspension input is donor 1 
When the user changes the linking so that: cell suspension input to donor 2
Then the changes will be submitted to the staging area upon export

#### assay became non-assay (This was made out of scope scenario from ebi-ait/dcp-ingest-central#680 )
2. Given 2 cell suspensions and 6 files, when cell suspension 1 is linked to the first 3 files and cell suspension 2 is linked to the next 3 files
When the user changes the process linking cell suspension 2 to 3 files so that the 3 files will now link to 3 files of cell suspension 1
Then the changes will be submitted to the staging area upon export. (No links.json file should contain the previous linking)

### II - has linked new metadata
3. Given 2 cell suspensions and 6 files, when cell suspension 1 is linked to 6 files
When the user adds a new process and changes the linking so that: cell suspension 1 linked to file 1,2,3 & cell suspension 2 linked to files 4,5,6
Then the changes will be submitted to the staging area upon export

4. Given 2 donors and cell suspension and cell suspension input is donor 1
When the user adds a new protocol and links the protocol to the process connecting cell suspension and donor 1
Then the changes will be submitted to the staging area upon export

### IV - has unlinked a deleted entity 
5. Given protocol and process, the protocol is linked to process
When the user removes the protocol
Then changes will be submitted to DCP, (The staging area shouldn't contain the protocol entity and the links.json should be updated and shouldn't refer to the removed protocol)

## Implementation notes

Per the decision that **during data import - delete everything for a project (all rows in terra) and reimport everything**, for Ingest, this means we should make sure the staging area only contains entities and links needed for reimport. Any deleted entities and links.json should be removed. (See [8th March meeting details](https://docs.google.com/document/d/1nTZDW9xQhgZfM6jQkBbK2O68CRvALPiJ8F-noiEvxTA/edit#heading=h.6ijg7tqg7h5k))

1. Ingest will also have to make sure that the links directory only contains all sets of links.json files containing the correct linking. Ingest could **clear the `links/` directory** for the project before every export.

2. Ingest will also need to delete any entities from the original staging area when entities are deleted in Ingest We could also **clear `metadata/` directory**   - they shouldn’t be reimported. 

3. (Added 10th March) We need to make sure that the **`data/` directory** should always have the complete set of data files in the Terra staging area. 

See  [Exporting updates - design doc](https://docs.google.com/document/d/1nTZDW9xQhgZfM6jQkBbK2O68CRvALPiJ8F-noiEvxTA/edit#) for more context about updates

DCP PR Spec changes: https://github.com/HumanCellAtlas/dcp2/pull/60

**Comment by aaclan-ebi** [comment]
Some new messages on this thread: https://embl-ebi-ait.slack.com/archives/C01360XN04S/p1646834044480059?thread_ts=1646223040.031019&cid=C01360XN04S

It feels like we can't proceed with this yet. The problem with the decision we made from the meeting last [8th March](https://docs.google.com/document/d/1nTZDW9xQhgZfM6jQkBbK2O68CRvALPiJ8F-noiEvxTA/edit#heading=h.6ijg7tqg7h5k) is that **we will always need to have the data files in the original Terra staging area**

1. If the Data Import team deletes them, how would we reexport data files to that staging area if they are already deleted in Ingest?

2. I asked the Data Import team if they can guarantee that we'll keep the data files in the staging area and they [commented](https://embl-ebi-ait.slack.com/archives/C01360XN04S/p1646834044480059?thread_ts=1646223040.031019&cid=C01360XN04S) they can. However, this is against the fact that the staging area should be a temporary location for files data import. (This is actually why they're called "staging" area.)

 I'll have another discussion with @amnonkhen today to discuss this and see how to proceed.

**Comment by Wkt8** [comment]
Needs clarification with downstream components which will not happen until the post-demo meeting with the DCP. @aaclan-ebi will set up a meeting before then.

**Comment by idazucchi** [comment]
@aaclan-ebi posted the PR for review on slack, no reviews yet

**Comment by ofanobilbao** [comment]
Hannes and Nathan have suggested changes but other people in the DCP need to review

**Comment by Wkt8** [comment]
Alegria to add some additional DCP component reviewers.

**Comment by gabsie** [comment]
@aaclan-ebi will involve @MightyAx / @jacobwindsor  when doing these changes to have the knowledge

**Comment by aaclan-ebi** [comment]
@amnonkhen @MightyAx @jacobwindsor PR for DCP spec changes https://github.com/HumanCellAtlas/dcp2/pull/60

**Comment by aaclan-ebi** [comment]
An issue was discovered in the scenario when there are multiple submissions in a project.
Exporting only the submission will export the entities for that submission and probably some entities from the previous submissions if linked to them. The new “update” staging area will not contain everything about the project. To fix this, we agreed on exporting the whole project when a submission is exported.

The idea for implementation: Update ingest core so that it will try to find all assays of the project instead of trying to only find the assays in a submission.

**Comment by MightyAx** [comment]
@aaclan-ebi  Remind them in the DCP demo?

**Comment by ami-day** [comment]
waiting for approval

**Comment by prabh-t** [comment]
@amnonkhen needs to check with the import team about the progress on this issue.

**Comment by ESapenaVentura** [comment]
@gabsie to add this to post-demo agenda: @amnonkhen to prep for discussion

**Comment by MightyAx** [comment]
I've prompted the dcp2 channel about this as it came up in the dcp demo, as a vital step in resolving the double pipes issue:
ebi-ait/hca-ebi-wrangler-central#809

**Comment by ESapenaVentura** [comment]
Waiting on people to review/approve PR 60 in DCP2 repo. Needs more research for now

**Comment by Wkt8** [comment]
Talk about this at the DCP2 demo. @idazucchi to put into the agenda.
